Feature/NSA-7979_implement_manage_console_access

AMcCartneyMTH opened the pull request:
Support Console - add ability to manage users roles within the Manage Console.

AMcCartneyMTH committed:
Initial commit with functioning implementation of manage console access role management.

AMcCartneyMTH committed:
Corrections for linting.

AMcCartneyMTH committed:
Fix to tests applied.

AMcCartneyMTH committed:
Refactored postManageConsoleRoles tests.

AMcCartneyMTH committed:
Refactored tests for readability and the removal of repitition.

alexplattdfe committed:
Support Console add user ability to manage Manage Console roles

alexplattdfe committed:
updated views removed comments.

alexplattdfe committed:
updated failing test, with neccessary mocks

AMcCartneyMTH committed:
Merge branch 'feature/NSA-7979_implement_manage_console_access' of https://github.com/DFE-Digital/login.dfe.support into feature/NSA-7979_implement_manage_console_access

AMcCartneyMTH committed:
Merge branch 'main' of https://github.com/DFE-Digital/login.dfe.support into feature/NSA-7979_implement_manage_console_access

alexplattdfe committed:
changed capitalisation on Manage console in summary view, and changed weighting of service name in manage console services view

alexplattdfe committed:
Merge branch 'feature/NSA-7979_implement_manage_console_access' of https://github.com/DFE-Digital/login.dfe.support into feature/NSA-7979_implement_manage_console_access

alexplattdfe committed:
Removed/commented the new layout from getManageConsoleServices.js and made the required changes to selectManageConsoleService.ejs so the table reflected the current  style

alexplattdfe committed:
updated success banner to current style

JaredAMarsh commented on src/app/users/getManageConsoleServices.js:
```diff
@@ -0,0 +1,19 @@
+  const services = await getAllServices();
+
+  sendResult(req, res, 'users/views/selectManageConsoleService', {
+    // layout: 'sharedViews/layoutNew.ejs', 
```
Was there an issue with the newLayout for why it wasn't used here? 😄 As it may be a little jarring flipping from old layout to new layout and back again 👍

alexplattdfe replied:
Thanks Jared, no issue with the new layout, more that the classes from the other existing table components in the app were not working with the new layout. Then the classes from the example table components from the gds website did not give a consistent look with the rest of the app.

JaredAMarsh commented on src/app/users/getManageConsoleRoles.js:
```diff
@@ -0,0 +1,69 @@
+
+const getManageConsoleRoles = async (req, res) => {
+  
+  const manage = await getServiceById('manage');
```
I'm quite happy for this additional call to be here, as the performance hit will be negligible, but if we're configuring the DfE org ID, adding this to the config as well may be useful 👍 If not now then down the line.

alexplattdfe replied:
Thanks Jared, the app now adds the manage console service if the user doesn't have it, and will only update if they do have the service.

JaredAMarsh commented on src/app/users/views/selectManageConsoleRoles.ejs:
```diff
@@ -0,0 +1,54 @@
+      <form method="post">
+        <input type="hidden" name="_csrf" value="<%=csrfToken%>" />
+          <div class="govuk-form-group">
+            <fieldset class="govuk-fieldset" aria-describedby="waste-hint" id="roleSelection">
```
Could we rename `waste-hint` to `role-hint` or similar? On this line and the one below it?

JaredAMarsh commented on src/app/users/postManageConsoleRoles.js:
```diff
@@ -0,0 +1,62 @@
+    if (!checkIfRolesChanged(currentRoles, newRoles)) {
+      return res.redirect(`/users/${req.params.uid}/manage-console-services`);
+    } else {
+      updateUserService(req.params.uid, manage.id, config.access.identifiers.departmentForEducation, newRoles, req.id);
```
🟥 Will this work if the user doesn't already have manage console access? [Update user service](https://github.com/DFE-Digital/login.dfe.access/blob/d1be2d5520982c8cba7f2ed2df2a5adf3b16653d/src/app/users/updateUserService.js#L62-L65) appears to return 404 if the user doesn't have it already 😄 

I can see we've made the change to only show these pages if a user already has manage access, but that'll mean a script will need to be run to set up that access initially 👍

alexplattdfe replied:
Hey Jared, thanks for the feedback - I did think the update would add the service if the user didnt have it, but thanks for pointing out that this isnt the case. I will add it in explicitly 😊

JaredAMarsh reviewed (changes requested):
All in all I can see a lot of positive changes in here, which covers most of the essence of the ticket, the key things I would like to know are:

- Why are we only updating manage console access now and not adding it if the user doesn't have it? 😄 We have the `addUserService` method from the access API we could call to add it if the user doesn't have it.
  - As things stand right now this will only update access, so for a user to get manage access in the first place it'll need to be added via script unless I'm missing something.
- Why isn't the new layout used in getManageConsoleServices? If it's an issue then that's okay, more of a curiosity question.

Overall, this is a great first ticket for you, and I know you've hit a bunch of roadblocks throughout but from where this started I can see a definite improvement in your code quality and understanding of the application 🌟

kruncher commented on .gitignore:
```diff
@@ -77,3 +77,6 @@ access.log
 \.idea/markdown-navigator\.xml
 
 \.idea/markdown-navigator/profiles_settings\.xml
+certs/PKI-ROOT-CA.cer
```
It would probably be better to exclude the entire `certs/` directory here instead of specific files to avoid inadvertently committing such a file if another is created for whatever reason.

alexplattdfe replied:
Thanks Lea 👍

alexplattdfe replied:
This has now been changed

kruncher commented on src/app/users/getManageConsoleRoles.js:
```diff
@@ -0,0 +1,69 @@
+  }
+};
+
+const addOrChangeManageConsoleServiceTitle = (userManageRoles, manageConsoleRoleIds) => {
```
You might want to consider something like the following:
```
const addOrChangeManageConsoleServiceTitle = (userManageRoles, manageConsoleRoleIds) => {
  const userManageConsoleRoleIds = userManageRoles.map(role => role.id);
  return manageConsoleRoleIds.some(roleId => userManageConsoleRoleIds.includes(roleId));
}
```

Also it's better to use `const` where possible over `let`.

alexplattdfe replied:
Thanks Lea. With this one I would like to return a boolean, not an array. I think this should do the trick?? 

`const addOrChangeManageConsoleServiceTitle = (userManageRoles, manageConsoleRoleIds) => {
  return userManageRoles.roles
    .map(role => role.id)
    .some(roleId => manageConsoleRoleIds.includes(roleId));
}`

kruncher replied:
Hi @alexplattdfe you should be able to achieve the above without .map since you can check `role.id` instead of mapping an array of roles to an array of roleIds within the `.some` predicate.

alexplattdfe replied:
Thanks for the feedback, i have now made these changes.

kruncher commented on src/app/users/getManageConsoleRoles.js:
```diff
@@ -0,0 +1,69 @@
+  const manageConsoleRolesForAllServices = await listRolesOfService(manage.id);
+  const manageConsoleRolesForSelectedService = manageConsoleRolesForAllServices.filter(service => service.code.split('_')[0] === req.params.sid);
+    
+  let manageConsoleRoleIds = [];
```
```
const manageConsoleRoleIds = manageConsoleRolesForSelectedService.map(service => service.id);
```

kruncher commented on src/app/users/utils.js:
```diff
@@ -213,8 +213,20 @@ const mapUserToSupportModel = (user, userFromSearch) => {
+      userServiceIds.push(arr[i].serviceId)
+    }
+
+    return userServiceIds.includes(manage.id)
```
Similar sort of thing:
```
const userServiceIds = arr.map(entry => entry.serviceId);
```
Although it seems like you are building an array just to search it; so perhaps even:
```
return arr.some(entry => entry.serviceId === manage.id);
```

alexplattdfe replied:
Thanks Lea, need to brush up on my array methods 👍

kruncher commented on src/app/users/utils.js:
```diff
@@ -358,9 +373,9 @@ const waitForIndexToUpdate = async (uid, updatedCheck) => {
+    return { id: 10000, description: "Approver" };
   }
-  return { id: 0, description: 'End user' };
+  return { id: 0, description: "End user" };
```
Why are the quotes being changed to double quotes here? I think our informal standard is single quotes.

alexplattdfe replied:
Changed back, I think i had prettier formatting the file I was working in at one point and that might formatted on save.

alexplattdfe committed:
Changes made in response to code review

AMcCartneyMTH committed:
Added mock for hasManageAccess

alexplattdfe committed:
added test for addUserService if the user does not have manage console access

AMcCartneyMTH committed:
Adding user to DfE organisation when they are being granted manage console access for the first time.

AMcCartneyMTH committed:
Adding filter on services which are currently assigned to a user.

AMcCartneyMTH committed:
Security trimming the services grid and tidying up the code.

alexplattdfe committed:
Security trimmed the list of services to the current user

alexplattdfe committed:
removed comments and added trailing semi colons

AMcCartneyMTH committed:
Adding filter on services which are currently assigned to a user.

Security trimming the services grid and tidying up the code.

Security trimmed the list of services to the current user

removed comments and added trailing semi colons

alexplattdfe committed:
Merge branch 'main' of https://github.com/DFE-Digital/login.dfe.support into feature/NSA-7979_implement_manage_console_access

AMcCartneyMTH committed:
Merge branch 'feature/NSA-7979_implement_manage_console_access' of https://github.com/DFE-Digital/login.dfe.support into feature/NSA-7979_implement_manage_console_access

JaredAMarsh reviewed (dismissed):
The changes since my last viewing appear to be addressed, the new layout has been added to almost all the pages where possible, and where the new GDS classes worked correctly 👍 As well as the recent changes add the service to users that don't have access already 🌟 

My only slight adjustment would be to bring the manage service ID into the config so we don't need to use `getServiceById` every time to get the ID of the service. But I would not call that a showstopper, I'd advise a ticket is made to address this down the line though 👍

alexplattdfe committed:
added guard clause around getManageConsoleServices

alexplattdfe committed:
reverted back to all services shown in list for manage console services view

alexplattdfe committed:
updated tests in getManageConsoleServices to reflect code changes

alexplattdfe committed:
additional guard logic for getSingleUserService where user is not mapped to any services

kruncher commented on test/app/users/getManageConsoleServices.test.js:
```diff
@@ -0,0 +1,183 @@
+
+    expect(getAllServices).toHaveBeenCalled();
+    expect(getAllServices).toReturnWith(
+      { "services": [{ "description": "Service for testing purposes", "id": "service1Id", "isExternalService": true, "isHiddenService": false, "isIdOnlyService": false, "isMigrated": false, "name": "Service 1", "relyingParty": {} }, { "description": "Service for testing purposes", "id": "service2Id", "isExternalService": true, "isHiddenService": false, "isIdOnlyService": false, "isMigrated": fals ...
```
This is a very long and hard to read line; could the formatting be improved on this one?

alexplattdfe replied:
Agreed, think the formatter must have been playing up on this.

kruncher commented on test/app/users/getManageConsoleServices.test.js:
```diff
@@ -0,0 +1,183 @@
+          relyingParty: {},
+        },
+        {
+          id: 'service5Id',
```
The parameters of this test service seem the same as the one above; is it necessary to have this many?

alexplattdfe replied:
Yep, totally. I think the test did need it at one point but ended up simplifying it.

kruncher commented on src/app/users/utils.js:
```diff
@@ -291,7 +299,7 @@ const getAllServicesForUserInOrg = async (
     id: service.serviceId,
     dateActivated: service.accessGrantedOn,
-    name: '',
+    name: "",
```
Should be single quotes to be consistent with existing source code.

alexplattdfe replied:
Thanks Lea, I did install prettier at one point and i think it's been playing around with my linter.

kruncher commented on src/app/users/utils.js:
```diff
@@ -213,8 +213,13 @@ const mapUserToSupportModel = (user, userFromSearch) => {
+
 const getUserDetailsById = async (uid, correlationId) => {
-  if (uid.startsWith('inv-')) {
+  if (uid.startsWith("inv-")) {
```
Should be single quotes to be consistent with existing source code.

kruncher commented on package.json:
```diff
@@ -39,6 +39,7 @@
     "express-session": "^1.17.3",
     "helmet": "^7.0.0",
     "ioredis": "^5.3.2",
+    "jest-offline": "^1.0.1",
```
Is this needed? the dependency `jest-offline` is not actively maintained (8 years old).

kruncher reviewed (changes requested):
I've done a partial code review and have made a couple of comments :)

JaredAMarsh reviewed (changes requested):
After adding/removing manage roles to myself in TEST, all appears good, but the changes @kruncher has mentioned are a good thing to resolve before approval 👍

alexplattdfe committed:
fixed formatting errors and removed duplicate code in tests

AMcCartneyMTH committed:
Corrected formatting against linter rules.

AMcCartneyMTH committed:
Added guard clause around getManageConsoleServices

Guard clause added to cover scenario when attempting to grant manage console access to a user who is not mapped to any services.

AMcCartneyMTH committed:
Merge branch 'feature/NSA-7979_implement_manage_console_access' of https://github.com/DFE-Digital/login.dfe.support into feature/NSA-7979_implement_manage_console_access

AMcCartneyMTH reviewed (commented):
Review to be approved by AM
